src/main/java/net/smartcosmos/dao/things/ThingDao.java

```diff
@@ -0,0 +1,126 @@
+package net.smartcosmos.dao.things;
+
+import net.smartcosmos.dto.things.ThingCreate;
+import net.smartcosmos.dto.things.ThingResponse;
+import net.smartcosmos.dto.things.ThingUpdate;
+
+import javax.validation.ConstraintViolationException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public interface ThingDao {
+
+    /**
+     * Creates a thing in the realm of a given account.
+     *
+     * @param tenantId the tenant ID
+     * @param thingCreate the thing to create
+     * @return an {@link ThingResponse} instance for the created thing
+     * @throws ConstraintViolationException if the {@link ThingCreate} violates constraints enforced by the persistence service
+     */
+    ThingResponse create(String tenantId, ThingCreate thingCreate) throws ConstraintViolationException;
+
+    /**
+     * Updates a thing in the realm of a given account.
+     *
+     * @param tenantId the tenant ID
+     * @param thingUpdate the thing to update
+     * @return an {@link ThingResponse} instance for the updated thing
+     * @throws ConstraintViolationException if the {@link ThingUpdate} violates constraints enforced by the persistence service
+     */
+    Optional<ThingResponse> update(String tenantId, ThingUpdate thingUpdate) throws ConstraintViolationException;
+
+    /**
+     * Finds a thing matching a specified URN in the realm of a given account.
+     *
+     * @param tenantId the tenant ID
+     * @param urn the thing URN
+     * @return the {@link ThingResponse} instance for the retrieved thing or {@code empty} if the thing does not exist
+     */
+    Optional<ThingResponse> findByUrn(String tenantId, String urn);
```

Comment:
We will use *findByTypeAndUrn*, because *urn* is more specific than *type*, and we put the more specific to the end.